server/src/main/java/org/elasticsearch/cluster/metadata/DataLifecycle.java

```diff
@@ -53,31 +57,43 @@ public class DataLifecycle implements SimpleDiffable<DataLifecycle>, ToXContentO
     public static final ConstructingObjectParser<DataLifecycle, Void> PARSER = new ConstructingObjectParser<>(
         "lifecycle",
         false,
-        (args, unused) -> new DataLifecycle((TimeValue) args[0])
+        (args, unused) -> new DataLifecycle((Retention) args[0], false)
     );
 
     static {
-        PARSER.declareField(
-            ConstructingObjectParser.optionalConstructorArg(),
-            (p, c) -> TimeValue.parseTimeValue(p.textOrNull(), DATA_RETENTION_FIELD.getPreferredName()),
-            DATA_RETENTION_FIELD,
-            ObjectParser.ValueType.STRING_OR_NULL
-        );
+        PARSER.declareField(ConstructingObjectParser.optionalConstructorArg(), (p, c) -> {
+            String value = p.textOrNull();
+            if (value == null) {
+                return Retention.NULL;
+            } else {
+                return new Retention(TimeValue.parseTimeValue(value, DATA_RETENTION_FIELD.getPreferredName()));
+            }
+        }, DATA_RETENTION_FIELD, ObjectParser.ValueType.STRING_OR_NULL);
     }
 
     public static boolean isEnabled() {
         return DLM_FEATURE_FLAG.isEnabled();
     }
 
     @Nullable
-    private final TimeValue dataRetention;
+    private final Retention dataRetention;
+    private final boolean nullified;
 
     public DataLifecycle() {
-        this.dataRetention = null;
+        this(null, false);
     }
 
     public DataLifecycle(@Nullable TimeValue dataRetention) {
+        this(new Retention(dataRetention), false);
+    }
+
+    public DataLifecycle(Retention dataRetention) {
+        this(dataRetention, false);
+    }
+
+    private DataLifecycle(@Nullable Retention dataRetention, boolean nullified) {
         this.dataRetention = dataRetention;
+        this.nullified = nullified;
```

Comment:
Maybe worth a validation check? `nullified && dataRetention != null` is an invalid state, right? You're not doing that anywhere, but someone in the future might do something dumb, and it's not obvious what would take precedence in that state.

Reply:
Done